apps/sunrise/app.js

```diff
@@ -14,333 +13,207 @@ function loadLocation () {
-  }
-  const x = pos;
-  const y = ypos(x);
-
   g.setColor(0.2, 0.2, 0);
   // wide glow
-  if (x > sunRiseX && x < sunSetX) {
-    g.fillCircle(x, y, r + 20);
+  if (pos > sunRiseX && pos < sunSetX) {
+    g.fillCircle(pos, ypos(pos * 2), r + 20);
     g.setColor(0.5, 0.5, 0);
   }
   // smol glow
-  g.fillCircle(x, y, r + 8);
-}
-
-function seaLevel (hour) {
-  // hour goes from 0 to 24
-  // to get the X we divide the screen in 24
-  return ypos(xfromTime(hour));
+  g.fillCircle(pos, ypos(pos * 2), r + 8);
 }
 
 function ypos (x) {
-  // offset, resulting in zenith being at the correct time
-  return oy + (32 * Math.sin(((x + sunRiseX - 12) / w) * 6.28 ));
+  return (h / 1.7) + (32 * Math.sin(-(2 * Math.PI * x / w
+                                    - (Math.PI * solarNoonX) / w)));
 }
 
-function xfromTime (t) {
-  return (w / 24) * t;
+function xFromTime (time) {
+  return Math.round((w / 24) * (time.getHours() + time.getMinutes() / 60));
 }
 
 function drawBall () {
-  const now = new Date();
-  if (frames < 1 && realTime) {
-    pos = xfromTime(now.getHours() + now.getMinutes() / 60);
-  }
-  const x = pos;
-  const y = ypos(x);
-
   // glow
-  if (x > sunRiseX && x < sunSetX) {
+  if (pos > sunRiseX && pos < sunSetX) {
     g.setColor(1, 1, 1);
   } else {
     g.setColor(0.5, 0.5, 0);
   }
-  g.fillCircle(x, y, r);
+  g.fillCircle(pos, ypos(pos * 2), r);
   g.setColor(1, 1, 0);
-  g.drawCircle(x, y, r);
+  g.drawCircle(pos, ypos(pos * 2), r);
 }
+
 function drawClock () {
-  const now = new Date();
+  let posTime;
 
-  let hours = 0.0;
-  let mins = 0.0;
   if (realTime) {
-    hours = now.getHours();
-    mins = now.getMinutes();
-  } else {
-    hours = 24 * (pos / w);
-    const nexth = 24 * 60 * (pos / w);
-    mins = 59 - ((24 * 60) - nexth) % 60;
-
-    // this prevents the displayed time to jump from 11:50 to 12:59 to 12:07
-    if (mins == 59) {
-      hours--;
-    }
-  }
+    posTime = now;
 
-  g.setFont('Vector', 30);
-  g.setColor(realTime, 1, 1);
-  g.drawString(formatAsTime(hours, mins), w / 1.9, 32);
-  // day-month
-  if (realTime) {
+    // day-month
     const mo = now.getMonth() + 1;
     const da = now.getDate();
     g.setFont('6x8', 2);
-    g.drawString('' + da + '/' + mo, 5, 30);
+    g.drawString(da + '/' + mo, 6, 30);
+  } else {
+    posTime = new Date(24 * 3600 * (pos / w) * 1000 +
+                       60 * now.getTimezoneOffset() * 1000);
   }
+  g.setFont('Vector', 30);
+  g.setColor(realTime, 1, 1);
+  g.drawString(Locale.time(posTime, 1), w / 1.9, 32);
+}
+
+function initDay () {
+  getAltitude();
+  sunRiseX = xFromTime(SunCalc.getTimes(now, lat, lon, alt).sunrise);
+  sunSetX = xFromTime(SunCalc.getTimes(now, lat, lon, alt).sunset);
+  solarNoonX = xFromTime(SunCalc.getTimes(now, lat, lon, alt).solarNoon);
+  sr = Locale.time(SunCalc.getTimes(now, lat, lon, alt).sunrise, 1);
+  ss = Locale.time(SunCalc.getTimes(now, lat, lon, alt).sunset, 1);
+  fillSineLUT();
```

Comment:
`fillSineLUT()` isn't defined and so crashes the app on my watch with:

```
Uncaught ReferenceError: "fillSineLUT" is not defined
 at line 160 col 3 in sunrise.app.js
  fillSineLUT();
  ^
in function "initDay" called from line 243 col 11 in sunrise.app.js
  initDay();
          ^
in function "main" called from line 249 col 6 in sunrise.app.js
main();
     ^
Uncaught Error: Unhandled promise rejection: Error: Unhandled promise rejection: undefined
in function "getAltitude" called from line 154 col 15 in sunrise.app.js
getAltitude();
            ^
in function "initDay" called from line 170 col 13 in sunrise.app.js
    initDay();
            ^
in function "renderScreen" called from line 210 col 16 in sunrise.app.js
  renderScreen();
               ^
in function called from system
 at line 160 col 3 in sunrise.app.js
  fillSineLUT();
  ^
in function "initDay" called from line 170 col 13 in sunrise.app.js
    initDay();
            ^
in function "renderScreen" called from line 210 col 16 in sunrise.app.js
  renderScreen();
               ^
in function called from system
```

Reply:
When I comment it out the app seems to work though :)

Reply:
... but the sun doesn't snap to the sine curve anymore. Maybe that can be tracked back to this?